File: source/java-sdk/src/main/java/com/aerospike/examples/transactionprocessing/TopTransactionsAcrossDcs.java
package com.aerospike.examples.transactionprocessing;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicLong;

import com.aerospike.client.sdk.Record;
import com.aerospike.client.sdk.Session;
import com.aerospike.client.sdk.TypedDataSet;
import com.aerospike.client.sdk.TypedKey;
import com.aerospike.client.sdk.TypedKeyList;
import com.aerospike.client.sdk.TypedRecordStream;
import com.aerospike.client.sdk.cdt.MapOrder;
import com.aerospike.examples.Async;
import com.aerospike.examples.UseCase;
import com.aerospike.examples.transactionprocessing.model.Account;
import com.aerospike.examples.transactionprocessing.model.Transaction;

/**
 * SDK port of the legacy {@code TopTransactionsAcrossDcs} (see ../../java). Transactions can
 * arrive at either of two simulated DCs; each account keeps the most recent {@code
 * MAX_TRANSACTIONS} transaction ids per DC in a key-ordered map ({@code txns_dc1}/{@code
 * txns_dc2}, keyed by a zero-padded {@code "timestamp-id"} string so map key order == time order),
 * trimmed on every write. Reading the account's overall top transactions means merging both DC
 * maps and taking the most recent entries across both.
 * <p/>
 * The legacy version merges the two DC maps and takes the top N in one round trip via a nested
 * conditional expression ({@code Exp.cond} + {@code MapExp.putItems} + {@code
 * MapExp.getByIndexRange}); an earlier pass at this port assumed the AEL equivalent wasn't
 * expressible (the same pre-canonical-reference conclusion later found wrong for {@link
 * com.aerospike.examples.timeseries.TimeSeriesDemo}'s device filter) and merged the two maps
 * client-side instead. Re-checked against the canonical AEL reference and it works - see {@link
 * #getTopResults} for the derivation, including the {@code let}-binding needed to select a range
 * on a merged map, and the {@code when}-gated fallback for when one or both DC bins don't exist
 * yet (an account with no transactions in a DC, common early in a run).
 */
public class TopTransactionsAcrossDcs implements UseCase {

    private static final String BIN_DC1 = "txns_dc1";
    private static final String BIN_DC2 = "txns_dc2";
    private static final long NUM_ACCOUNTS = 1_000L;
    private static final int SIMULATION_DAYS = 30;
    private static final int MAX_TRANSACTIONS = 50;
    private static final long RUNTIME_SECS = 25L;

    private static final String[] DESCRIPTIONS = {"Grocery", "Gas station", "Online purchase", "ATM withdrawal", "Restaurant"};

    @Override
    public String getName() {
        return "Top 50 Transaction Across DCs";
    }

    @Override
    public String getDescription() {
        return "Find the top 50 transactions for an account. Transactions can arrive at either of 2 DCs at any point in time. "
                + "Transaction shipping from the remote DC is about 100ms. Transactions have a unique id but 2 transaction can arrive for the same "
                + "account at the same time.";
    }

    @Override
    public String getReference() {
        return "https://github.com/aerospike-examples/use-case-cookbook/blob/main/UseCases/top-transactions-across-dcs.md";
    }

    private final TypedDataSet<Account> accounts =
            TypedDataSet.of(System.getProperty("demo.namespace", "test"), "uccb_account", Account.class);
    private final TypedDataSet<Transaction> transactions =
            TypedDataSet.of(System.getProperty("demo.namespace", "test"), "uccb_txn", Transaction.class);

    public TypedKey<Account> getAccountKey(String id) {
        return accounts.id(id);
    }

    public TypedKey<Transaction> getTransactionKey(String id) {
        return transactions.id(id);
    }

    private String formMapKey(Transaction txn) {
        return String.format("%013d-%8s", txn.getTimestamp(), txn.getId());
    }

    @Override
    public void setup(Session session) throws Exception {
        session.truncate(accounts);
        session.truncate(transactions);

        System.out.printf("Generating %,d accounts%n", NUM_ACCOUNTS);
        for (long i = 1; i <= NUM_ACCOUNTS; i++) {
            session.upsert(accounts).object(new Account("acct-" + i)).execute();
        }
    }

    private Transaction randomTransaction(long counter) {
        String id = "txn-" + counter + "-" + ThreadLocalRandom.current().nextInt(1_000_000);
        String accountId = "acct-" + (ThreadLocalRandom.current().nextInt((int) NUM_ACCOUNTS) + 1);
        int amount = ThreadLocalRandom.current().nextInt(1, 100_000);
        Transaction.Status status = Transaction.Status.values()[ThreadLocalRandom.current().nextInt(Transaction.Status.values().length)];
        String desc = DESCRIPTIONS[ThreadLocalRandom.current().nextInt(DESCRIPTIONS.length)];
        String approvalCode = Long.toHexString(ThreadLocalRandom.current().nextLong());
        return new Transaction(id, 0L, amount, desc, status, null, approvalCode, accountId);
    }

    @Override
    public void run(Session session) throws Exception {
        Async.runFor(Duration.ofSeconds(RUNTIME_SECS), async -> {
            AtomicLong totalTxns = new AtomicLong();
            AtomicLong txnCounter = new AtomicLong();

            System.out.printf("Starting simulating time for %d seconds%n", RUNTIME_SECS);

            // We want the transactions to be roughly ordered to simulate real traffic patterns. Given the
            // execution should be short, we simulate the number of days of transactions we want over our run duration.
            async.useVirtualTime()
                    .elapse(Duration.ofDays(SIMULATION_DAYS))
                    .in(Duration.ofSeconds(RUNTIME_SECS))
                    .withPriorOffsetOf(Duration.ofDays(SIMULATION_DAYS))
                    .startingNow();

            System.out.println("Starting at: " + async.virtualDate());

            async.periodic(Duration.ofMillis(2500), () -> showTopTransactionsForAccount1(session, async, totalTxns));

            async.continuous(-1, () -> {
                Transaction txn = randomTransaction(txnCounter.incrementAndGet());
                String binName;
                if (async.rand().nextBoolean()) {
                    txn.setTimestamp(async.virtualTimeWithVariance(-15, 10));
                    binName = BIN_DC1;
                }
                else {
                    txn.setTimestamp(async.virtualTimeWithVariance(-100, -10));
                    binName = BIN_DC2;
                }
                txn.setOrigin(binName);
                session.upsert(transactions).object(txn).execute();

                // Insert the new transaction into the account's DC map, and trim that map to the most recent MAX_TRANSACTIONS.
                session.upsert(getAccountKey(txn.getAccountId()))
                        .bin(binName).onMapKey(formMapKey(txn), MapOrder.KEY_ORDERED).upsert(txn.getId())
                        .bin(binName).onMapIndexRange(-MAX_TRANSACTIONS).removeAllOthers()
                        .execute();
                totalTxns.incrementAndGet();
            });
        });
    }

    private void showTopTransactionsForAccount1(Session session, Async async, AtomicLong totalTxns) {
        System.out.printf("%s: %,d transactions generated%n", async.virtualDate(), totalTxns.get());
        long now = System.currentTimeMillis();
        List<Transaction> topResults = getTopResults(session, MAX_TRANSACTIONS, "acct-1");
        long time = System.currentTimeMillis() - now;
        for (int i = 0; i < topResults.size(); i++) {
            Transaction txn = topResults.get(i);
            System.out.printf("%4d: %10s %8s %10s  %s  $%d%n",
                    i + 1, txn.getId(), txn.getAccountId(), txn.getOrigin(), new Date(txn.getTimestamp()), txn.getAmount());
        }
        System.out.printf("%d transaction(s) retrieved in %,dms%n%n", topResults.size(), time);
    }

    /**
     * Retrieves the most recent transactions for an account across both DC maps, merged and
     * sorted by their "timestamp-id" map key. Over-fetches by a few entries since a transaction's
     * map entry can arrive slightly before the transaction record itself.
     * <p/>
     * Merge-plus-top-N is a single AEL read: {@code putItems} merges {@code txns_dc2} into {@code
     * txns_dc1} (key-ordered, so the merge result is already sorted); the merged map can't be
     * selector-ranged directly ({@code $.a.putItems($.b).{-N:}} is a parse error - {@code
     * putItems()} is a path write terminal, so the chain ends there), but binding it to a {@code
     * let} variable and re-navigating from {@code (${var})} (canonical reference §4.2's rule for
     * continuing a path after a parenthesised expression) works. A {@code when} wraps the whole
     * thing to handle a DC bin not existing yet - {@code $.dc1.putItems($.dc2)} throws if either
     * bin is absent (bin navigation is strict by default, §4.1), which happens for any account
     * that hasn't received a transaction in one or both DCs.
     */
    @SuppressWarnings("unchecked")
    public List<Transaction> getTopResults(Session session, int count, String accountId) {
        int countToUse = count + 3;
        String ael = String.format(
                "when ("
                        + "$.%1$s.exists() and $.%2$s.exists() => (let (merged = $.%1$s.putItems($.%2$s)) then ((${merged}).{-%3$d:})), "
                        + "$.%1$s.exists() => $.%1$s.{-%3$d:}, "
                        + "$.%2$s.exists() => $.%2$s.{-%3$d:}, "
                        + "default => []"
                        + ")",
                BIN_DC1, BIN_DC2, countToUse);
        Record record = session.query(getAccountKey(accountId)).bin("top").selectFrom(ael).execute().getFirstRecord();
        if (record == null) {
            return List.of();
        }
        List<?> topAscending = record.getList("top");

        List<String> txnIds = new ArrayList<>();
        for (int i = topAscending.size() - 1; i >= 0; i--) {
            txnIds.add((String) topAscending.get(i));
        }
        if (txnIds.isEmpty()) {
            // No transactions have landed for this account yet (e.g. a display tick racing ahead
            // of the generator right at startup) - a batch query requires at least one key.
            return List.of();
        }

        TypedKeyList<Transaction> keys = new TypedKeyList<>();
        txnIds.forEach(id -> keys.add(getTransactionKey(id)));

        try (TypedRecordStream<Transaction> stream = session.query(keys).execute()) {
            List<Transaction> txns = stream.toObjectList();
            return txns.size() > count ? new ArrayList<>(txns.subList(0, count)) : txns;
        }
    }
}

File: source/python-sdk/usecasecookbook/transactionprocessing/top_transactions_across_dcs.py
"""Port of ../../java-sdk's TopTransactionsAcrossDcs (itself an SDK port of ../../java).

Transactions can arrive at either of two simulated DCs; each account keeps the most
recent ``MAX_TRANSACTIONS`` transaction ids per DC in a key-ordered map (``txns_dc1``/
``txns_dc2``, keyed by a zero-padded ``"timestamp-id"`` string so map key order == time
order), trimmed on every write via the native CDT builder (``on_map_key(...,
create_type=...).set_to(...)`` then ``on_map_index_range(-N).remove_all_others()`` - this
SDK's equivalent of ../../java-sdk's ``.onMapKey(...).upsert(...)``/
``.onMapIndexRange(-N).removeAllOthers()``). Reading an account's overall top transactions
means merging both DC maps and taking the most recent entries across both - see
:meth:`TopTransactionsAcrossDcs.get_top_results` for why that merge is done client-side
here rather than as a single AEL read the way ../../java-sdk does it.
"""

import random
import threading
import time
from datetime import datetime, timedelta
from typing import Dict, List

from aerospike_async import MapOrder
from aerospike_sdk import DataSet, SyncSession

from usecasecookbook import config
from usecasecookbook.async_util import Async
from usecasecookbook.transactionprocessing.model import Account, Transaction
from usecasecookbook.use_case import UseCase

BIN_DC1 = "txns_dc1"
BIN_DC2 = "txns_dc2"
NUM_ACCOUNTS = 1_000
SIMULATION_DAYS = 30
MAX_TRANSACTIONS = 50
RUNTIME_SECS = 25.0

DESCRIPTIONS = ["Grocery", "Gas station", "Online purchase", "ATM withdrawal", "Restaurant"]
STATUSES = ["APPROVED", "DENIED", "FRAUD"]

# Set names must match ../../java-sdk exactly - uccb_account is intentionally shared with
# manytomany's own (unrelated) Account model, mirroring a pre-existing collision on the
# Java SDK side rather than introducing a different name here.
ACCOUNTS = DataSet.of(config.NAMESPACE, "uccb_account")
TRANSACTIONS = DataSet.of(config.NAMESPACE, "uccb_txn")


class _Counter:
    """A plain int isn't safe to increment from multiple ``Async`` threads at once -
    this is the same role as ../../java-sdk's ``AtomicLong``.
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._value = 0

    def increment(self) -> int:
        with self._lock:
            self._value += 1
            return self._value

    def get(self) -> int:
        with self._lock:
            return self._value


def _form_map_key(txn: Transaction) -> str:
    return f"{txn.timestamp:013d}-{txn.id:>8s}"


def _random_transaction(counter: int) -> Transaction:
    txn_id = f"txn-{counter}-{random.randint(0, 999_999)}"
    account_id = f"acct-{random.randint(1, NUM_ACCOUNTS)}"
    amount = random.randint(1, 99_999)
    status = random.choice(STATUSES)
    desc = random.choice(DESCRIPTIONS)
    approval_code = format(random.getrandbits(64), "x")
    return Transaction(txn_id, 0, amount, desc, status, None, approval_code, account_id)


class TopTransactionsAcrossDcs(UseCase):
    def get_name(self) -> str:
        return "Top 50 Transaction Across DCs"

    def get_description(self) -> str:
        return (
            "Find the top 50 transactions for an account. Transactions can arrive at either of 2 DCs at any point "
            "in time. Transaction shipping from the remote DC is about 100ms. Transactions have a unique id but 2 "
            "transaction can arrive for the same account at the same time."
        )

    def get_reference(self) -> str:
        return "https://github.com/aerospike-examples/use-case-cookbook/blob/main/UseCases/top-transactions-across-dcs.md"

    def setup(self, session: SyncSession) -> None:
        session.truncate(ACCOUNTS)
        session.truncate(TRANSACTIONS)

        print(f"Generating {NUM_ACCOUNTS:,} accounts")
        for i in range(1, NUM_ACCOUNTS + 1):
            account = Account(f"acct-{i}")
            session.upsert(ACCOUNTS.id(account.id)).put(account.to_bins()).execute()

    def run(self, session: SyncSession) -> None:
        def main_thread(async_runner: Async) -> None:
            total_txns = _Counter()
            txn_counter = _Counter()

            print(f"Starting simulating time for {RUNTIME_SECS:.0f} seconds")

            # We want the transactions to be roughly ordered to simulate real traffic patterns.
            # Given the run should be short, simulate SIMULATION_DAYS of transactions over
            # RUNTIME_SECS of wall-clock time, starting SIMULATION_DAYS in the (virtual) past.
            async_runner.use_virtual_time(
                timedelta(days=SIMULATION_DAYS),
                timedelta(seconds=RUNTIME_SECS),
                start_offset=timedelta(days=SIMULATION_DAYS),
            )

            print(f"Starting at: {async_runner.virtual_date()}")

            async_runner.periodic(
                2.5,
                lambda: self._show_top_transactions_for_account1(session, async_runner, total_txns),
            )

            def _generate() -> None:
                txn = _random_transaction(txn_counter.increment())
                if random.random() < 0.5:
                    # ~15ms either side of "now" - the local DC.
                    txn.timestamp = int(async_runner.virtual_time_with_variance(-0.015, 0.010) * 1000)
                    bin_name = BIN_DC1
                else:
                    # 10-100ms behind "now" - the remote DC, per get_description()'s ~100ms shipping time.
                    txn.timestamp = int(async_runner.virtual_time_with_variance(-0.100, -0.010) * 1000)
                    bin_name = BIN_DC2
                txn.origin = bin_name
                session.upsert(TRANSACTIONS.id(txn.id)).put(txn.to_bins()).execute()

                # Insert the new transaction into the account's DC map, and trim that map to
                # the most recent MAX_TRANSACTIONS entries.
                map_key = _form_map_key(txn)
                session.upsert(ACCOUNTS.id(txn.account_id)) \
                    .bin(bin_name).on_map_key(map_key, create_type=MapOrder.KEY_ORDERED).set_to(txn.id) \
                    .bin(bin_name).on_map_index_range(-MAX_TRANSACTIONS).remove_all_others() \
                    .execute()
                total_txns.increment()

            async_runner.continuous(_generate, number_of_copies=-1)

        Async.run_for(RUNTIME_SECS, main_thread)

    def _show_top_transactions_for_account1(
        self, session: SyncSession, async_runner: Async, total_txns: _Counter,
    ) -> None:
        print(f"{async_runner.virtual_date()}: {total_txns.get():,} transactions generated")
        start = time.monotonic()
        top_results = self.get_top_results(session, MAX_TRANSACTIONS, "acct-1")
        elapsed_ms = (time.monotonic() - start) * 1000
        for i, txn in enumerate(top_results, start=1):
            print(
                f"{i:4d}: {txn.id:>10s} {txn.account_id:>8s} {txn.origin:>10s}  "
                f"{datetime.fromtimestamp(txn.timestamp / 1000)}  ${txn.amount}"
            )
        print(f"{len(top_results)} transaction(s) retrieved in {elapsed_ms:,.0f}ms\n")

    def get_top_results(self, session: SyncSession, count: int, account_id: str) -> List[Transaction]:
        """Return an account's most recent transactions across both DC maps, newest first.

        ../../java-sdk merges the two per-DC maps and takes the top N as a single AEL read
        (``let (merged = $.dc1.putItems($.dc2)) then ((${merged}).{-N:})``, wrapped in a
        ``when`` for a DC bin that doesn't exist yet). This SDK's AEL grammar has no
        equivalent map-merge path function - confirmed via the ANTLR grammar (there is no
        ``putItems`` token anywhere in ``Condition.g4``) and empirically: running that exact
        expression through ``select_from`` raises ``AelParseException: line 1:28 no viable
        alternative at input 'let(merged=$.dc1.putItems('``. So the merge is done
        client-side here instead: read the account record (both DC maps come back as plain
        dicts keyed by the same zero-padded "timestamp-id" string used to write them, so
        string order == chronological order), merge the two dicts, sort the combined keys
        descending, and take the top few. Over-fetches by a few entries (same as
        ../../java-sdk) since a transaction's map entry can arrive slightly before the
        transaction record itself is written.
        """
        stream = session.query(ACCOUNTS.id(account_id)).execute()
        record = None
        for row in stream:
            if row.is_ok:
                record = row.record
        stream.close()
        if record is None:
            return []

        merged: Dict[str, str] = {}
        merged.update(record.bins.get(BIN_DC1) or {})
        merged.update(record.bins.get(BIN_DC2) or {})
        if not merged:
            return []

        count_to_use = count + 3
        top_keys = sorted(merged.keys(), reverse=True)[:count_to_use]
        txn_ids = [merged[k] for k in top_keys]

        keys = [TRANSACTIONS.id(tid) for tid in txn_ids]
        txn_stream = session.query(keys).execute()
        txns: List[Transaction] = []
        for row in txn_stream:
            if row.is_ok and row.record is not None:
                txns.append(Transaction.from_bins(row.record.bins))
        txn_stream.close()
        return txns[:count]
